<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Enrichment analysis (hypergeometric test) for the fungal DEGs</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h1>Enrichment analysis (hypergeometric test) for the fungal DEGs</h1>

<p>This code runs the enrichment_analysis function for the fungal DEGs
as a sanity check to see if the function works as expected and gives us 
the same results as using any other program for enrichment analysis.</p>

<pre><code class="r">library(tidyverse)
</code></pre>

<h1>Initial loading and restructuring the data files</h1>

<p>Set your working directory to the folder with the csv files
Read the file with Ophio gene annotations
[Replace the following file with your organism&#39;s gene annotation; ensure the column names GO, PFAM, etc are the same]</p>

<pre><code class="r">all_genes &lt;- read.csv(&quot;./FullBlast_EC05_RNAseq_orignal_copy_26Aug19.csv&quot;, header = T, stringsAsFactors = FALSE)
</code></pre>

<pre><code class="r">names(all_genes)

</code></pre>

<p>Select only the columns that we need.</p>

<pre><code class="r">all_genes &lt;- all_genes[,c(&quot;arb2_gene&quot;,&quot;GO&quot;,&quot;PFAM&quot;,&quot;signalp&quot;,&quot;TMHMM&quot;)]
head(all_genes)
##      arb2_gene                                                            GO
## 1 Ophcf2|00001 GO:0003824|catalytic activity; GO:0003674|molecular_function;
## 2 Ophcf2|00001 GO:0003824|catalytic activity; GO:0003674|molecular_function;
## 3 Ophcf2|00001 GO:0003824|catalytic activity; GO:0003674|molecular_function;
## 4 Ophcf2|00002                                                          &lt;NA&gt;
## 5 Ophcf2|00002                                                          &lt;NA&gt;
## 6 Ophcf2|00002                                                          &lt;NA&gt;
##                                PFAM signalp TMHMM
## 1              PF00501|AMP-binding;    &lt;NA&gt;  &lt;NA&gt;
## 2              PF00501|AMP-binding;    &lt;NA&gt;  &lt;NA&gt;
## 3              PF00501|AMP-binding;    &lt;NA&gt;  &lt;NA&gt;
## 4 PF13415|Kelch_3; PF13418|Kelch_4;    &lt;NA&gt;  &lt;NA&gt;
## 5 PF13415|Kelch_3; PF13418|Kelch_4;    &lt;NA&gt;  &lt;NA&gt;
## 6 PF13415|Kelch_3; PF13418|Kelch_4;    &lt;NA&gt;  &lt;NA&gt;
</code></pre>

<p>There seem to be duplicate rows for the same gene. Keep only one row per gene.</p>

<pre><code class="r">all_genes &lt;- dplyr::distinct((all_genes))
head(all_genes)
##      arb2_gene
## 1 Ophcf2|00001
## 2 Ophcf2|00002
## 3 Ophcf2|00003
## 4 Ophcf2|00004
## 5 Ophcf2|00005
## 6 Ophcf2|00006
##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      GO
## 1                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         GO:0003824|catalytic activity; GO:0003674|molecular_function;
## 2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  &lt;NA&gt;
## 3                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0003824|catalytic activity; GO:0051920|peroxiredoxin activity; GO:0055114|oxidation-reduction process; GO:0003674|molecular_function; GO:0016491|oxidoreductase activity; GO:0008150|biological_process; GO:0016684|oxidoreductase activity, acting on peroxide as acceptor; GO:0016209|antioxidant activity; GO:0008152|metabolic process;
## 4                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        GO:0016298|lipase activity; GO:0003824|catalytic activity; GO:0016787|hydrolase activity; GO:0052689|carboxylic ester hydrolase activity; GO:0003674|molecular_function; GO:0004806|triglyceride lipase activity; GO:0016788|hydrolase activity, acting on ester bonds; GO:0008150|biological_process; GO:0071704|organic substance metabolic process; GO:0006629|lipid metabolic process; GO:0044238|primary metabolic process; GO:0008152|metabolic process;
## 5 GO:0006694|steroid biosynthetic process; GO:0016614|oxidoreductase activity, acting on CH-OH group of donors; GO:0003674|molecular_function; GO:0071704|organic substance metabolic process; GO:0006629|lipid metabolic process; GO:0003854|3-beta-hydroxy-delta5-steroid dehydrogenase activity; GO:0016229|steroid dehydrogenase activity; GO:0055114|oxidation-reduction process; GO:0003824|catalytic activity; GO:0008610|lipid biosynthetic process; GO:0016491|oxidoreductase activity; GO:0044238|primary metabolic process; GO:0033764|steroid dehydrogenase activity, acting on the CH-OH group of donors, NAD or NADP as acceptor; GO:0008202|steroid metabolic process; GO:0009058|biosynthetic process; GO:1901362|organic cyclic compound biosynthetic process; GO:0008150|biological_process; GO:1901360|organic cyclic compound metabolic process; GO:1901576|organic substance biosynthetic process; GO:0008152|metabolic process; GO:0016616|oxidoreductase activity, acting on the CH-OH group of donors, NAD or NADP as acceptor;
## 6                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      GO:0003824|catalytic activity; GO:0140096|catalytic activity, acting on a protein; GO:0016787|hydrolase activity; GO:0005488|binding; GO:0008270|zinc ion binding; GO:0046872|metal ion binding; GO:0003674|molecular_function; GO:0008233|peptidase activity; GO:0043167|ion binding; GO:0043169|cation binding; GO:0046914|transition metal ion binding; GO:0070011|peptidase activity, acting on L-amino acid peptides; GO:0008237|metallopeptidase activity;
##                                                             PFAM signalp TMHMM
## 1                                           PF00501|AMP-binding;    &lt;NA&gt;  &lt;NA&gt;
## 2                              PF13415|Kelch_3; PF13418|Kelch_4;    &lt;NA&gt;  &lt;NA&gt;
## 3         PF00578|AhpC-TSA; PF08534|Redoxin; PF10417|1-cysPrx_C;    &lt;NA&gt;  &lt;NA&gt;
## 4                              PF01734|Patatin; PF11815|DUF3336;    &lt;NA&gt;  &lt;NA&gt;
## 5       PF05368|NmrA; PF13460|NAD_binding_10; PF01073|3Beta_HSD;    &lt;NA&gt;  &lt;NA&gt;
## 6 PF17900|Peptidase_M1_N; PF01433|Peptidase_M1; PF11838|ERAP1_C;    &lt;NA&gt;  &lt;NA&gt;
</code></pre>

<h1>The following code performs GO enrichment only.</h1>

<p>Repeat the same protocol for any other enrichment (PFAM, signalp, etc)</p>

<p>Is there any NAs in the GO column, meaning, are there fungal genes with no GO annotations?</p>

<pre><code class="r">summary(is.na(as.factor(all_genes$GO)))
##    Mode   FALSE    TRUE 
## logical    3515    3940
</code></pre>

<p>Now, let&#39;s replace the NAs in GOs and PFAMs with the term &ldquo;no_annot&rdquo;</p>

<pre><code class="r">all_genes[is.na(all_genes)] &lt;- &quot;no_annot&quot;
</code></pre>

<p>Check for NAs again</p>

<pre><code class="r">summary(is.na(as.factor(all_genes$GO)))
##    Mode   FALSE 
## logical    7455
</code></pre>

<p>There is no more NAs since we replaced them with the category &ldquo;no annot&rdquo;.
Now, let&#39;s flatten the file (aim: for each gene, one GO term per line)</p>

<pre><code class="r">all_genes_gos &lt;- all_genes %&gt;%
  dplyr::mutate(go_split = str_split(GO, &quot;; &quot;)) %&gt;% 
  unnest() %&gt;% 
  #dplyr::select(-GO) %&gt;%
  separate(go_split, c(&quot;GO&quot;,&quot;GO_desc&quot;), sep = &quot;([\\|])&quot;, extra = &quot;drop&quot;) %&gt;% 
  dplyr::select(gene_name = &quot;arb2_gene&quot;, GO, GO_desc)
</code></pre>

<p>Let&#39;s take a look to see if it worked.</p>

<pre><code class="r">head(all_genes_gos)
## # A tibble: 6 x 3
##   gene_name    GO         GO_desc                    
##   &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;                      
## 1 Ophcf2|00001 GO:0003824 catalytic activity         
## 2 Ophcf2|00001 GO:0003674 molecular_function;        
## 3 Ophcf2|00002 no_annot   &lt;NA&gt;                       
## 4 Ophcf2|00003 GO:0003824 catalytic activity         
## 5 Ophcf2|00003 GO:0051920 peroxiredoxin activity     
## 6 Ophcf2|00003 GO:0055114 oxidation-reduction process

unique(all_genes_gos$gene_name) %&gt;% 
  length()  ## sanity check to see all starting genes have made it to the flattened file
## [1] 7455
</code></pre>

<p>How many unique GO terms are there in the Ophio genome?</p>

<pre><code class="r">unique(all_genes_gos$GO) %&gt;% 
  length() 
## [1] 2400
</code></pre>

<h1>Now letâ€™s make the function that will do the enrichment analysis</h1>

<pre><code class="r">enrichment_analysis &lt;- function(background, geneset) {

  # this is going to be your test geneset
  genes &lt;- geneset

  # background genes to run enrichment against 
  df.whole &lt;- background # (the flattened &#39;all_genes_gos&#39; file above)

  go_to_desc &lt;- dplyr::distinct(as.data.frame(background[-1]))

  # subset the background file to keep only genes in the test set
  df.test &lt;- background %&gt;% 
    filter(gene_name %in% genes)

  ## all unique GO terms in the background data set
  annot_terms &lt;- unique(df.whole[[2]])  

  ## Create an empty list to save results from the enrichment test
  df.list &lt;- list()


  for (i in 1:length(annot_terms)) {

    # assign the annotation term to test for enrichment
    annot &lt;- annot_terms[i]

    # number of genes in the test set
    k &lt;- length(genes)

    # number of genes that have the annotation term in the background geneset
    m &lt;- df.whole %&gt;% 
      filter(GO == annot) %&gt;% 
      nrow()

    # number of genes that DON&#39;T have the annotation term in the background geneset
    n &lt;- length(unique(df.whole[[1]])) - m

    # number of genes that have the annotation term in the test geneset
    x &lt;- df.test %&gt;% 
      filter(GO == annot) %&gt;% 
      nrow()

    ### Perform the hypergeometric test
    pval &lt;- dhyper(x, m, n, k, log=F)

    # Save the results into the list
    df.list[[i]] &lt;- data.frame(GO = annot_terms[i],
                               x = x,
                               k=k,
                               m=m,
                               n=n,
                               tot_genes = sum(m,n),
                               pVal = pval)


  }


  # Row-bind the results for all tested annotation terms
  df.enriched &lt;- bind_rows(df.list, .id = &quot;column_label&quot;) %&gt;% 
    select(-column_label) %&gt;% 
    arrange(pVal) %&gt;% 
    ## Adjust the p-values for multiple hypothesis testing (BH-correction)
    mutate(adj_pVal = p.adjust(pVal, &quot;BH&quot;)) %&gt;% 
    ## Filter to keep annotation terms that have adjusted p-value &lt; 0.1
    filter(adj_pVal &lt; 0.1) %&gt;%
    ## add a column with the descriptions for each of the GO terms
    left_join(go_to_desc, by=&quot;GO&quot;)


  # Return the results as a data.frame
  return(df.enriched);

}
</code></pre>

<h1>Read a test geneset</h1>

<p>read file</p>

<pre><code class="r">ophio_rnaseq &lt;- read.csv(&quot;./enrichtest_fungal.csv&quot;, stringsAsFactors = F, header = T)
</code></pre>

<p>I just need the gene names (first column), let&#39;s only keep that.</p>

<pre><code class="r">ophio_rnaseq &lt;- ophio_rnaseq[,1, drop = T]   ## drop = T is necessary for the function to work. 
head(ophio_rnaseq)
## [1] &quot;Ophcf2|00036&quot; &quot;Ophcf2|00037&quot; &quot;Ophcf2|00145&quot; &quot;Ophcf2|00188&quot; &quot;Ophcf2|00402&quot; &quot;Ophcf2|00450&quot;
</code></pre>

<h1>Let&#39;s run the enrichment analysis</h1>

<pre><code class="r">ophio_genes_enrich &lt;- enrichment_analysis(background = all_genes_gos, geneset = ophio_rnaseq)
</code></pre>

<p>What are top most hits?</p>

<pre><code class="r">knitr::kable(head(ophio_genes_enrich))
</code></pre>

<table><thead>
<tr>
<th align="left">GO</th>
<th align="right">x</th>
<th align="right">k</th>
<th align="right">m</th>
<th align="right">n</th>
<th align="right">tot_genes</th>
<th align="right">pVal</th>
<th align="right">adj_pVal</th>
<th align="left">GO_desc</th>
</tr>
</thead><tbody>
<tr>
<td align="left">GO:0006270</td>
<td align="right">5</td>
<td align="right">168</td>
<td align="right">7</td>
<td align="right">7448</td>
<td align="right">7455</td>
<td align="right">0.0000001</td>
<td align="right">0.0002643</td>
<td align="left">DNA replication initiation</td>
</tr>
<tr>
<td align="left">GO:0044421</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular region part</td>
</tr>
<tr>
<td align="left">GO:0090729</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">toxin activity</td>
</tr>
<tr>
<td align="left">GO:0005615</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular space;</td>
</tr>
<tr>
<td align="left">GO:0005615</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular space</td>
</tr>
<tr>
<td align="left">GO:0009405</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">35</td>
<td align="right">7420</td>
<td align="right">7455</td>
<td align="right">0.0001027</td>
<td align="right">0.0493157</td>
<td align="left">pathogenesis</td>
</tr>
</tbody></table>

<p>What about the tail?</p>

<pre><code class="r">knitr::kable(tail(ophio_genes_enrich))
</code></pre>

<table><thead>
<tr>
<th align="left"></th>
<th align="left">GO</th>
<th align="right">x</th>
<th align="right">k</th>
<th align="right">m</th>
<th align="right">n</th>
<th align="right">tot_genes</th>
<th align="right">pVal</th>
<th align="right">adj_pVal</th>
<th align="left">GO_desc</th>
</tr>
</thead><tbody>
<tr>
<td align="left">3</td>
<td align="left">GO:0090729</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">toxin activity</td>
</tr>
<tr>
<td align="left">4</td>
<td align="left">GO:0005615</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular space;</td>
</tr>
<tr>
<td align="left">5</td>
<td align="left">GO:0005615</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular space</td>
</tr>
<tr>
<td align="left">6</td>
<td align="left">GO:0009405</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">35</td>
<td align="right">7420</td>
<td align="right">7455</td>
<td align="right">0.0001027</td>
<td align="right">0.0493157</td>
<td align="left">pathogenesis</td>
</tr>
<tr>
<td align="left">7</td>
<td align="left">GO:0044419</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">38</td>
<td align="right">7417</td>
<td align="right">7455</td>
<td align="right">0.0001635</td>
<td align="right">0.0560722</td>
<td align="left">interspecies interaction between organisms</td>
</tr>
<tr>
<td align="left">8</td>
<td align="left">GO:0051704</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">38</td>
<td align="right">7417</td>
<td align="right">7455</td>
<td align="right">0.0001635</td>
<td align="right">0.0560722</td>
<td align="left">multi-organism process</td>
</tr>
</tbody></table>

<p>Let&#39;s look at only the GO terms that have a BH-adjusted pVal &lt; 0.06</p>

<pre><code class="r">enriched_ophio &lt;- ophio_genes_enrich %&gt;% 
  filter(adj_pVal &lt; 0.06)

knitr::kable(enriched_ophio)
</code></pre>

<table><thead>
<tr>
<th align="left">GO</th>
<th align="right">x</th>
<th align="right">k</th>
<th align="right">m</th>
<th align="right">n</th>
<th align="right">tot_genes</th>
<th align="right">pVal</th>
<th align="right">adj_pVal</th>
<th align="left">GO_desc</th>
</tr>
</thead><tbody>
<tr>
<td align="left">GO:0006270</td>
<td align="right">5</td>
<td align="right">168</td>
<td align="right">7</td>
<td align="right">7448</td>
<td align="right">7455</td>
<td align="right">0.0000001</td>
<td align="right">0.0002643</td>
<td align="left">DNA replication initiation</td>
</tr>
<tr>
<td align="left">GO:0044421</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular region part</td>
</tr>
<tr>
<td align="left">GO:0090729</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">toxin activity</td>
</tr>
<tr>
<td align="left">GO:0005615</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular space;</td>
</tr>
<tr>
<td align="left">GO:0005615</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">34</td>
<td align="right">7421</td>
<td align="right">7455</td>
<td align="right">0.0000870</td>
<td align="right">0.0493157</td>
<td align="left">extracellular space</td>
</tr>
<tr>
<td align="left">GO:0009405</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">35</td>
<td align="right">7420</td>
<td align="right">7455</td>
<td align="right">0.0001027</td>
<td align="right">0.0493157</td>
<td align="left">pathogenesis</td>
</tr>
<tr>
<td align="left">GO:0044419</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">38</td>
<td align="right">7417</td>
<td align="right">7455</td>
<td align="right">0.0001635</td>
<td align="right">0.0560722</td>
<td align="left">interspecies interaction between organisms</td>
</tr>
<tr>
<td align="left">GO:0051704</td>
<td align="right">6</td>
<td align="right">168</td>
<td align="right">38</td>
<td align="right">7417</td>
<td align="right">7455</td>
<td align="right">0.0001635</td>
<td align="right">0.0560722</td>
<td align="left">multi-organism process</td>
</tr>
</tbody></table>

<h1>Converting the R script to a markdown file.</h1>

<p>Type the following in the R console:
knitr::spin(&#39;./tutorial/enrichment_fungal_rnaseq_IW.R&#39;)</p>

</body>

</html>
